### PR TITLE
add stack trace to device skip

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118113
* __->__ #118112

Log stack trace of offending cpu use if it causes a disabling of cudagraphs. Also refactoring disable_cudagraphs: bool, and disable_cudagraphs_reason: str -> Optional[str].

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/cudagraph_utils.py [hunk 1]
```diff
@@ -54,3 +54,35 @@ def check_for_mutation(
...
+
+
+def get_use_stack_trace(node) -> Optional[str]:
+    for use in node.users:
+        if stack_trace := use.meta.get("stack_trace", None):
+            return stack_trace
+    return None
+
+
+def check_multiple_devices(
```
Comment (bdhirsh) [inline torch/_inductor/cudagraph_utils.py:66]: probably a nit but: maybe rename to `check_multiple_devices_or_any_cpu_nodes`? IIRC, you can have a graph that's 100% cpu (single-device) but we want this fn to trip and not use cudagraphs

**Review by bdhirsh — APPROVED** [review]
lgtm, nits

File: torch/_inductor/cudagraph_utils.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/cudagraph_utils.py:66]: Sure will do

**Pull request closed without merging**